plugins/chat/app/services/chat/list_channel_messages.rb

```diff
@@ -0,0 +1,131 @@
+      attribute :target_message_id, :integer # (optional)
+      attribute :direction, :string # (optional)
+      attribute :fetch_from_last_read, :boolean # (optional)
+      attribute :target_date, :string # (optional)
+
+      validates :direction,
+                inclusion: {
+                  in: Chat::MessagesQuery::VALID_DIRECTIONS,
+                },
+                allow_nil: true
+    end
+
+    private
+
+    def fetch_channel(contract:, **)
+      ::Chat::Channel.strict_loading.includes(:chatable).find_by(id: contract.channel_id)
+    end
+
+    def fetch_eventual_membership(channel:, guardian:, **)
+      context.membership = channel.membership_for(guardian.user)
+    end
+
+    def enabled_threads?(channel:, **)
+      context.enabled_threads =
+        SiteSetting.enable_experimental_chat_threaded_discussions && channel.threading_enabled
+    end
+
+    def can_view_channel(guardian:, channel:, **)
+      guardian.can_preview_chat_channel?(channel)
+    end
+
+    def determine_target_message_id(contract:, **)
+      contract.target_message_id =
+        context.membership&.last_read_message_id if contract.fetch_from_last_read
+    end
+
+    def target_message_exists(contract:, channel:, guardian:, **)
+      return true if contract.target_message_id.blank?
+      target_message =
+        Chat::Message.with_deleted.find_by(id: contract.target_message_id, chat_channel: channel)
+      return false if target_message.blank?
+      return true if !target_message.trashed?
+      target_message.user_id == guardian.user.id || guardian.is_staff?
+    end
+
+    def fetch_messages(channel:, guardian:, contract:, **)
+      messages_data =
+        ::Chat::MessagesQuery.call(
+          channel: channel,
+          guardian: guardian,
+          target_message_id: contract.target_message_id,
+          page_size: contract.page_size || Chat::MessagesQuery::MAX_PAGE_SIZE,
+          direction: contract.direction,
+          target_date: contract.target_date,
+        )
+
+      context.can_load_more_past = messages_data[:can_load_more_past]
+      context.can_load_more_future = messages_data[:can_load_more_future]
+
+      context.messages = [
+        messages_data[:messages],
+        messages_data[:past_messages]&.reverse,
+        messages_data[:target_message],
+        messages_data[:future_messages],
+      ].flatten.compact
+    end
+
+    def fetch_tracking(guardian:, enabled_threads:, messages:, **)
+      context.tracking = {}
+
+      return if !enabled_threads
+
+      thread_ids = messages.map(&:thread_id).compact.uniq
+      if thread_ids.present?
+        context.tracking =
+          ::Chat::TrackingStateReportQuery.call(
+            guardian: guardian,
+            thread_ids: thread_ids,
+            include_threads: true,
+          )
+      end
+    end
+
+    def update_membership_last_viewed_at(guardian:, **)
+      context.membership&.update!(last_viewed_at: Time.zone.now)
+    end
+  end
```

Comment:
We lost the thread participants here, so in the UI threads that have multiple participants are not showing their avatars:

![image](https://github.com/discourse/discourse/assets/920448/13d598c4-b785-48c2-9db6-eae53ffb75b2)

The old view builder had this:

```ruby
def fetch_thread_participants(threads:, **)
  context.thread_participants =
    ::Chat::ThreadParticipantQuery.call(thread_ids: threads.map(&:id))
end
```

Which it passed into `ViewSerializer`

Reply:
The participants show up once you send a message in the thread, because of this:

https://github.com/discourse/discourse/blob/20ec7ac1741d3931a65bdc015c35895a18e12164/plugins/chat/app/services/chat/publisher.rb#L86-L101